### PR TITLE
[Lint] Pylint

### PR DESCRIPTION
**Pull request by jermainewang**
## Description

This PR enables pylint check for our python codes. Related to #290 .

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR

## Changes
- [x] Pylint for most of the python codes under python/dgl.
- [x] Fix some bugs in inplace write. @ylfdq1118 
- [x] Simplify the logic in frame with slice index type. This could be done because `utils.Index` now supports slice data. @BarclayII is there any model using slice on the frame?

Following codes are not enabled for lint check.
* `dgl.backend` : TBD.
* `dgl.contrib` : @zheng-da could you help?
* `dgl.data` : need help from the owner.
* `dgl.nn` : this is likely to be changed in the future.

One could check the lint results offline by (requires cpplint and pylint):
```
bash tests/scripts/task_lint.sh
```

We also need to put this in a developer guide in our doc.

### CODE REVIEW

**Comment by jermainewang** [comment]
@ylfdq1118 and @BarclayII could you have a look at the changes in frame.py? Other changes are mainly style fixes.

File: python/dgl/frame.py [hunk 1]
```diff
@@ -115,22 +130,21 @@ def update(self, idx, feats, inplace):
...
 
         if inplace:
+            idx = idx.tousertensor(F.context(self.data))
             F.scatter_row_inplace(self.data, idx, feats)
+        elif idx.slice_data() is not None:
+            # for contiguous indices narrow+concat is usually faster than scatter row
+            slc = idx.slice_data()
+            part1 = F.narrow_row(self.data, 0, slc.start)
+            part2 = feats
+            part3 = F.narrow_row(self.data, slc.stop, len(self))
```
[the remark this answers is not in the document]
Comment (BarclayII) [reply python/dgl/frame.py]: Yeah I noticed that as well.  In fact this problem exists in the old code as well.  I fixed it on my unified test branch but maybe you can fix that as well and I merge later?

File: python/dgl/batched_graph.py [hunk 2]
```diff
@@ -152,7 +152,7 @@ def _init_attrs(attrs, mode):
             elif is_all(attrs):
                 attrs = set()
                 # Check if at least a graph has mode items and associated features.
-                for i in range(len(graph_list)):
+                for i, g in enumerate(graph_list):
                     g = graph_list[i]
```
[the remark this answers is not in the document]
Comment (jermainewang) [reply python/dgl/batched_graph.py]: Ah, that's not needed.

File: python/dgl/function/base.py [hunk 3]
```diff
@@ -1,26 +1,32 @@
...
-        raise NotImplementedError
-
     @property
     def name(self):
         """Return the name of this builtin function."""
         raise NotImplementedError
 
-class BundledFunction(object):
+class BundledFunction(BuiltinFunction):
+    """A builtin function that bundles multiple functions.
```
[the remark this answers is not in the document]
Comment (jermainewang) [reply python/dgl/function/base.py]: Got it.

File: python/dgl/frame.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jermainewang) [reply python/dgl/frame.py]: @BarclayII I will leave it to you.

[the remark this answers is not in the document]
Comment (jermainewang) [reply python/dgl/frame.py]: Also, is this a problem of MX or ours?

[the remark this answers is not in the document]
Comment (ylfdq1118) [reply python/dgl/frame.py]: My feeling is this is a problem with MX. And actually, MXNet's behavior looks weird to me. For `nd.slice`, as long as `begin=end`, `nd.slice` will return a tensor with the same shape of the original tensor.

For example, if we have `a=nd.normal(shape=(10, 5))`, then `nd.slice(a, begin=0, end=0)` will return `a`, but `nd.slice(a, begin=2, end=2)` will start slicing from the 2nd row and takes out 10 rows, which goes out boundary without any errors. So the last two rows of return tensor has random values... And if argument `begin` goes out of range, e.g. `begin=10`, it throws out exception.

[the remark this answers is not in the document]
Comment (jermainewang) [reply python/dgl/frame.py]: Raise an issue in mx? @zheng-da please help in tracking it.

**Pull request merged**